tests/core/test_ensemble.py

```diff
@@ -1,566 +1,212 @@
+    ),
 )
-def test_is_not_in_training_data(
-    policy_name: Text, confidence: Optional[float], not_in_training_data: bool
+def test_default_combine_predictions(
+    default_ensemble: DefaultPolicyPredictionEnsemble,
+    predictions_and_winner_idx: Tuple[List[PolicyPrediction2], int],
+    last_action_was_action_listen: bool,
 ):
-    assert (
-        SimplePolicyEnsemble.is_not_in_training_data(policy_name, confidence)
-        == not_in_training_data
-    )
-
+    predictions, winner_idx = predictions_and_winner_idx
 
-def test_rule_action_wins_over_action_unlikely_intent(
-    monkeypatch: MonkeyPatch,
-    tmp_path: Path,
-    unexpected_intent_policy_agent: Agent,
-    moodbot_domain: Domain,
-):
-    # The original training data consists of a rule for `goodbye` intent.
-    # We monkey-patch UnexpecTEDIntentPolicy to always predict action_unlikely_intent
-    # if last user intent was goodbye. The predicted action from ensemble
-    # should be utter_goodbye and not action_unlikely_intent.
-    monkeypatch.setattr(
-        UnexpecTEDIntentPolicy,
-        "predict_action_probabilities",
-        _action_unlikely_intent_for("goodbye"),
-    )
+    # add mandatory and optional events to every prediction
+    for prediction in predictions:
+        prediction.events = [ActionExecuted(action_name=prediction.policy_name)]
+        prediction.optional_events = [
+            ActionExecuted(action_name=f"optional-{prediction.policy_name}")
+        ]
 
-    tracker = DialogueStateTracker.from_events(
-        "rule triggering tracker",
-        evts=[
-            ActionExecuted(ACTION_LISTEN_NAME),
-            UserUttered(text="goodbye", intent={"name": "goodbye"}),
-        ],
+    evts = (
+        [ActionExecuted(action_name=ACTION_LISTEN_NAME)]
+        if last_action_was_action_listen
+        else []
     )
-    policy_ensemble = unexpected_intent_policy_agent.policy_ensemble
-    prediction = policy_ensemble.probabilities_using_best_policy(
-        tracker, moodbot_domain, NaturalLanguageInterpreter()
-    )
-
-    test_utils.assert_predicted_action(prediction, moodbot_domain, "utter_goodbye")
+    tracker = DialogueStateTracker.from_events(sender_id="arbitrary", evts=evts)
 
+    best_prediction = default_ensemble.combine_predictions(predictions, tracker)
 
-def test_ensemble_prevents_multiple_action_unlikely_intents(
-    monkeypatch: MonkeyPatch,
-    tmp_path: Path,
-    unexpected_intent_policy_agent: Agent,
-    moodbot_domain: Domain,
-):
-    monkeypatch.setattr(
-        UnexpecTEDIntentPolicy,
-        "predict_action_probabilities",
-        _action_unlikely_intent_for("greet"),
-    )
-
-    tracker = DialogueStateTracker.from_events(
-        "rule triggering tracker",
-        evts=[
-            ActionExecuted(ACTION_LISTEN_NAME),
-            UserUttered(text="hello", intent={"name": "greet"}),
-            ActionExecuted(ACTION_UNLIKELY_INTENT_NAME),
-        ],
+    # expected events
+    expected_events = set(
```

Comment:
this is somewhat tautological testing, isn't it? How about hardcoding the expected events instead of computing them in the test?

Reply:
I'm not sure what you mean here. The "computed" events are included in the `best_prediction`. I can move the "expected" stuff before the "combine" step to emphasise that?

Reply:
(and maybe rename `winner_idx` to `expected winner idx`)

Reply:
done, better now?